### PR TITLE
fix(hashline-edit): harden deduplication, validation, and add benchmark suite

### PR DESCRIPTION
**Pull request by minpeter**
## Summary

Fixes several hashline-edit edge cases that caused silent failures, and adds a comprehensive 46-test benchmark suite to validate the tool against real LLM behavior.

### Hashline-edit fixes (main)

- **Canonicalize anchors in dedupe keys** — whitespace variants of the same anchor (`1#AB` vs ` 1#AB `) now correctly deduplicate instead of producing duplicate edits
- **Enforce `lines` field as required in schema** — prevents models from sending edits without content, which previously caused silent no-ops
- **Reject anchored edits on missing files** — only unanchored `append`/`prepend` can create new files; anchored edits on non-existent files now return a clear error
- **Reorder delete/rename validation** — delete mode validation now runs before normalization, preventing edge cases where invalid delete edits would slip through
- **5 new regression tests** covering all fixed paths

### Benchmark suite (sub)

Ported from [code-editing-agent](https://github.com/can1357/code-editing-agent) — standalone benchmark that tests the hashline-edit tool against real LLM outputs.

```
benchmarks/
├── package.json          # Own deps: ai@^6.0.94, zod@^4.1.0, @friendliai/ai-provider
├── bun.lock              # Lock file for reproducibility
├── headless.ts           # Standalone agent loop (190 LOC)
├── test-edit-ops.ts      # 21 basic tests
├── test-edge-cases.ts    # 25 edge cases
└── test-multi-model.ts   # Multi-model comparison runner
```

**How to run:**
```bash
cd benchmarks && bun install
FRIENDLI_TOKEN=<token> bun run bench:basic   # 21 basic tests
FRIENDLI_TOKEN=<token> bun run bench:edge    # 25 edge cases
```

## Results (Minimax M2.5 via FriendliAI)

| Suite | Pass | Fail | Rate |
|-------|------|------|------|
| Basic Edit Ops (21) | 21 | 0 | **100%** |
| Edge Cases (25) | 25 | 0 | **100%** |
| **Combined** | **46** | **0** | **100%** |

### Hashline-edit
- `src/tools/hashline-edit/edit-deduplication.ts` — canonical anchor in dedupe keys
- `src/tools/hashline-edit/hashline-edit-executor.ts` — reorder delete validation, reject anchored edits on missing files
- `src/tools/hashline-edit/tools.ts` — enforce `lines` required in schema
- `src/tools/hashline-edit/tool-description.ts` — update `lines` description
- `src/tools/hashline-edit/validation.ts` — export `normalizeLineRef`
- `src/tools/hashline-edit/tools.test.ts` — 4 new regression tests
- `src/tools/hashline-edit/edit-operations.test.ts` — 1 new dedupe test
- `assets/oh-my-opencode.schema.json` — schema update

### Benchmark (new)
- `benchmarks/package.json`, `benchmarks/bun.lock`, `benchmarks/headless.ts`
- `benchmarks/test-edit-ops.ts`, `benchmarks/test-edge-cases.ts`, `benchmarks/test-multi-model.ts`

### CODE REVIEW

File: benchmarks/test-multi-model.ts [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+      console.log(`    ${icon} ${t.name}`);
+    }
+  }
+
+  console.log();
+
+  // Overall
+  const totalModels = allResults.length;
+  const perfectModels = allResults.filter(
+    (r) => r.totalPassed === r.totalTests
```
[the remark this answers is not in the document]
Comment (minpeter) [reply benchmarks/test-multi-model.ts]: Fixed in [`8fb5949`](https://github.com/code-yeongyu/oh-my-opencode/commit/8fb5949ac653057602892a4bf6745627d3e6b524). Thanks for catching this.

File: benchmarks/test-multi-model.ts [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+    console.log();
+  }
+
+  // ── Summary Table ──────────────────────────────────────────
+  console.log(`${BOLD}═══ Summary ═══${RESET}\n`);
+
+  // Per-model results
+  for (const r of allResults) {
+    const timeStr = `${(r.durationMs / 1000).toFixed(0)}s`;
+    const color = r.totalPassed === r.totalTests ? GREEN : r.totalPassed > 0 ? YELLOW : RED;
```
[the remark this answers is not in the document]
Comment (minpeter) [reply benchmarks/test-multi-model.ts]: Fixed in [`8fb5949`](https://github.com/code-yeongyu/oh-my-opencode/commit/8fb5949ac653057602892a4bf6745627d3e6b524). Thanks for catching this.

File: benchmarks/test-multi-model.ts [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+  // { id: "zai-org/GLM-5", short: "GLM-5" },            // masked: API 503
+  { id: "zai-org/GLM-4.7", short: "GLM-4.7" },
+];
+
+// ── CLI args ──────────────────────────────────────────────────
+let perModelTimeoutSec = 900; // 15 min default per model (5 tests)
+const rawArgs = process.argv.slice(2);
+for (let i = 0; i < rawArgs.length; i++) {
+  if (rawArgs[i] === "--timeout" && i + 1 < rawArgs.length) {
+    perModelTimeoutSec = Number.parseInt(rawArgs[i + 1], 10);
```
[the remark this answers is not in the document]
Comment (minpeter) [reply benchmarks/test-multi-model.ts]: Fixed in [`8fb5949`](https://github.com/code-yeongyu/oh-my-opencode/commit/8fb5949ac653057602892a4bf6745627d3e6b524). Thanks for catching this.

File: benchmarks/test-edge-cases.ts [hunk 4]
```diff
@@ -0,0 +1,1117 @@
...
+          reason: `expected 3 lines, got ${lines.length}`,
+        };
+      }
+      if (lines[0] !== "alpha" || lines[2] !== "omega") {
+        return {
+          passed: false,
+          reason: "non-target lines changed unexpectedly",
+        };
+      }
+      if (lines[1].trim() !== "middle-content") {
```
[the remark this answers is not in the document]
Comment (minpeter) [reply benchmarks/test-edge-cases.ts]: Fixed in [`8fb5949`](https://github.com/code-yeongyu/oh-my-opencode/commit/8fb5949ac653057602892a4bf6745627d3e6b524). Thanks for catching this.

File: benchmarks/test-edge-cases.ts [hunk 5]
```diff
@@ -0,0 +1,1117 @@
...
+  tc: TestCase,
+  testDir: string
+): Promise<{
+  passed: boolean;
+  editCalls: number;
+  editSuccesses: number;
+  duration: number;
+}> {
+  const testFile = join(testDir, tc.fileName);
+  writeFileSync(testFile, tc.fileContent, "utf-8");
```
[the remark this answers is not in the document]
Comment (minpeter) [reply benchmarks/test-edge-cases.ts]: Fixed in [`8fb5949`](https://github.com/code-yeongyu/oh-my-opencode/commit/8fb5949ac653057602892a4bf6745627d3e6b524). Thanks for catching this.

File: benchmarks/test-edge-cases.ts [hunk 6]
```diff
@@ -0,0 +1,1117 @@
...
+ *   bun run scripts/test-headless-edit-edge-cases.ts [-m <model>] [--provider <provider>]
+ */
+
+import { spawn } from "node:child_process";
+import { mkdirSync, readFileSync, rmSync, writeFileSync } from "node:fs";
+import { tmpdir } from "node:os";
+import { join, resolve } from "node:path";
+
+// ── CLI arg passthrough ───────────────────────────────────────
+const extraArgs: string[] = [];
```
[the remark this answers is not in the document]
Comment (minpeter) [reply benchmarks/test-edge-cases.ts:18]: By design — benchmark files are intentionally standalone (no shared imports) so they can be copied/run independently without wiring up a shared module. The duplication is ~15 lines of CLI parsing, acceptable for standalone scripts.

File: benchmarks/headless.ts [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+            tool_name: part.toolName,
+            tool_input: part.args,
+            model: modelId,
+          })
+          break
+        case "tool-result":
+          emit({
+            type: "tool_result",
+            tool_call_id: part.toolCallId,
+            output: typeof part.result === "string" ? part.result : JSON.stringify(part.result),
```
[the remark this answers is not in the document]
Comment (minpeter) [reply benchmarks/headless.ts]: Fixed in [`8fb5949`](https://github.com/code-yeongyu/oh-my-opencode/commit/8fb5949ac653057602892a4bf6745627d3e6b524). Thanks for catching this.

File: benchmarks/test-edge-cases.ts [hunk 8]
```diff
@@ -0,0 +1,1117 @@
...
+const CYAN = "\x1b[36m";
+const RESET = "\x1b[0m";
+
+const pass = (msg: string) => console.log(`  ${GREEN}✓${RESET} ${msg}`);
+const fail = (msg: string) => console.log(`  ${RED}✗${RESET} ${msg}`);
+const info = (msg: string) => console.log(`  ${DIM}${msg}${RESET}`);
+const warn = (msg: string) => console.log(`  ${YELLOW}⚠${RESET} ${msg}`);
+
+// ── Test case definition ─────────────────────────────────────
+interface TestCase {
```
[the remark this answers is not in the document]
Comment (minpeter) [reply benchmarks/test-edge-cases.ts:51]: Same rationale as CLI duplication above — benchmark test files are standalone by design. Each file should run in isolation without shared dependencies. The TestCase interface is simple enough (~10 lines) that extracting it would add coupling without meaningful benefit.

File: benchmarks/test-edge-cases.ts [hunk 9]
```diff
@@ -0,0 +1,1117 @@
...
+      `Replace line 2 using edit_file with edits: [{ op: 'replace', pos: '<line2 anchor>', lines: ['${"L".repeat(180)}'] }].`,
+      "Expected line 2 to be exactly 180 characters.",
+    ].join(" "),
+    validate: (content) => {
+      const expected = "L".repeat(180);
+      const lines = content.replace(/\r/g, "").trimEnd().split("\n");
+      if (!lines[1]) {
+        return { passed: false, reason: "line 2 is missing" };
+      }
+      if (Math.abs(lines[1].length - 180) > 2) {
```
[the remark this answers is not in the document]
Comment (minpeter) [reply benchmarks/test-edge-cases.ts:467]: The ±2 char tolerance is intentional. LLMs sometimes add/drop a trailing space or minor punctuation in long lines (~180 chars). The benchmark tests whether the model can handle long lines at all, not exact character reproduction. Strict matching would cause false negatives from trivial whitespace differences.

File: benchmarks/headless.ts [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+          })
```
[the remark this answers is not in the document]
Comment (minpeter) [reply benchmarks/headless.ts:162]: Good catch. Fixed in [`8fb5949`](https://github.com/code-yeongyu/oh-my-opencode/commit/8fb5949ac653057602892a4bf6745627d3e6b524) — `tool_result` events now include an `error` field when output starts with `"Error:"`.

File: benchmarks/test-edge-cases.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (minpeter) [reply benchmarks/test-edge-cases.ts]: Good catch. Fixed in [`8fb5949`](https://github.com/code-yeongyu/oh-my-opencode/commit/8fb5949ac653057602892a4bf6745627d3e6b524) — added `skipFileCreate` flag for test #7 so it actually tests new-file creation via append.

File: benchmarks/test-edge-cases.ts [hunk 11]
```diff
@@ -0,0 +1,1121 @@
...
+  tc: TestCase,
+  testDir: string
+): Promise<{
+  passed: boolean;
+  editCalls: number;
+  editSuccesses: number;
+  duration: number;
+}> {
+  const testFile = join(testDir, tc.fileName);
+  if (!tc.skipFileCreate) {
```
[the remark this answers is not in the document]
Comment (minpeter) [reply benchmarks/test-edge-cases.ts:912]: This is a false positive. The `readFileSync` call at line 1008 is already wrapped in a try-catch (lines 1007-1016) that returns `passed: false` if the file does not exist. The missing-file path is properly handled — no unguarded read occurs.
